server/src/main/java/io/crate/replication/logical/action/TransportCreatePublicationAction.java

```diff
@@ -103,18 +115,59 @@ protected ClusterState execute(ClusterState currentState,
                 var newMetadata = PublicationsMetadata.newInstance(oldMetadata);
+                var publication = new Publication(request.owner(), request.isForAllTables(), request.tables());
                 newMetadata.publications().put(
-                    request.name(),
-                    new Publication(request.owner(), request.isForAllTables(), request.tables())
+                    publicationName,
+                    publication
                 );
                 assert !newMetadata.equals(oldMetadata) : "must not be equal to guarantee the cluster change action";
                 mdBuilder.putCustom(PublicationsMetadata.TYPE, newMetadata);
+                List<RelationName> tablesToUpdate;
+                if (request.isForAllTables()) {
+                    tablesToUpdate = resolveRelationsNames(publication, currentState, user);
+                } else {
+                    tablesToUpdate = request.tables();
+                }
+                applyPublicationSetting(indexNameExpressionResolver,
+                                        tablesToUpdate,
+                                        currentState,
+                                        mdBuilder,
+                                        setting -> setting.add(publicationName));
 
                 return ClusterState.builder(currentState).metadata(mdBuilder).build();
             }
         };
     }
 
+    static void applyPublicationSetting(IndexNameExpressionResolver indexNameExpressionResolver,
+                                      List<RelationName> tables,
+                                      ClusterState currentState,
+                                      Metadata.Builder mdBuilder,
+                                      Consumer<List<String>> consumer) {
+        for (var relationName : tables) {
+            var concreteIndices = indexNameExpressionResolver.concreteIndexNames(
+                currentState,
+                IndicesOptions.lenientExpandOpen(),
+                relationName.indexNameOrAlias()
+            );
+            for (var indexName : concreteIndices) {
+                IndexMetadata indexMetadata = currentState.metadata().index(indexName);
+                assert indexMetadata != null : "Cannot resolve indexMetadata for relation=" + relationName;
+                var settings = indexMetadata.getSettings();
+                var publicationNames = REPLICATION_PUBLICATION_NAMES.get(settings);
+                consumer.accept(publicationNames);
+                var settingsBuilder = Settings.builder().put(settings);
+                settingsBuilder.putList(REPLICATION_PUBLICATION_NAMES.getKey(), publicationNames);
+                mdBuilder.put(
+                    IndexMetadata
+                        .builder(indexMetadata)
+                        .settingsVersion(1 + indexMetadata.getSettingsVersion())
+                        .settings(settingsBuilder)
+                );
+            }
+        }
+    }
```

Comment:
Could we instead compute this information on a need basis instead of duplicating it within the cluster state?

Or what's the motivation behind this? Seems like this redundancy will only make it harder to ensure things are kept in sync?

Reply:
I agree on this, i would also prefer the cluster state directly. Currently it is consistent with `LogicalReplicationSettings.REPLICATION_SUBSCRIPTION_NAME`.